sympy/combinatorics/tests/test_homomorphisms.py

```diff
@@ -83,11 +84,18 @@ def test_isomorphisms():
     assert check
     assert T(b*a*b**-1*a**-1*b**-1) == Permutation(0, 2, 3)
     assert T(b*a*b*a**-1*b**-1) == Permutation(0, 3, 2)
+    # Find all possible epimorphisms.
+    list_hom = find_homomorphism(G, H, injective=True, all=True)
```

Comment:
It would be better to do this test for `G` and `H` that are not isomorphic.
Additionally, a test of either injective or surjective (but not both at the same time) homomorphisms with `compute=False` could be good. For example, you could take `G` and `H` where `G.order()` doesn't divide `H.order()`, or where `G` is abelian and `H` isn't, and assert that there isn't an injective homomorphism.

Another thing I've noticed, is that you redefine free groups in the second and third tests when you could just reuse `F` and `E`